Feat: LIP-30 - Triggearble Withdrawals (#82)
lidofinance/lido-improvement-proposals: LIPS/lip-31.md

Reviewer: what am I missing now that we should say:
- Lido facilitates exists now through VEBO softly enforcing them with penalties for the delays
- This LIP suggests to use the same mechanism optimistically but enforce exits instead of time-based penalties

Probably can put it in the simple summary section and a separate backwards compat section

Reviewer: please rename the file to lip-30

---

> lip: 31

Reviewer: ```suggestion
lip: 30
```

title: Triggerable Withdrawals Framework
status: WIP
author: Raman Siamionau, Evgeniy Pirogov
discussions-to: TBD
created: 2025-05-21
updated: 2025-06-11
---

## Simple Summary

> This proposal outlines implementing the Triggerable Withdrawal framework within the Lido protocol to enable permissionless, secure, and verifiable validator exits via the Execution Layer - enhancing the protocol’s fault tolerance, reducing trust assumptions, and paving the way for truly permissionless staking in Lido.

Reviewer: ```suggestion
This proposal outlines implementing the Triggerable Withdrawals framework within the Lido protocol to enable permissionless, secure, and verifiable validator exits initiated from the Execution Layer — enhancing the protocol’s fault tolerance, reducing trust assumptions, and enhancing the rails for the permissionless staking in Lido.
```

## Abstract

> The **Triggerable Withdrawals (TW)** mechanism is a critical extension of Lido’s architecture, enabling the initiation of validator exits without requiring the involvement of a Node Operator. TW is based on [EIP-7002](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7002.md), which addresses a key issue in delegated staking. Previously, stakers had to rely on the goodwill of Node Operators - either to pre-sign an exit message or to agree to process it in the future. This limitation is now removed: any party with access to a validator’s withdrawal credentials can initiate its exit directly via the Execution Layer. 

Reviewer: ```suggestion
The **Triggerable Withdrawals (TW)** mechanism is a critical extension of Lido’s architecture, enabling the initiation of validator exits (and partial withdrawals in future) from the EL side without relying on the involvement of a Node Operator from the CL side. TW is based on [EIP-7002](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7002.md), which addresses one of the key long-standing issues in delegated staking on Ethereum. Previously, stakers had to rely on the goodwill of Node Operators - either to pre-sign an exit message or to agree to process it in the future. This limitation is now removed: any party with access to a validator’s withdrawal credentials (non-`0x00` types) can initiate its exit directly via the Execution Layer. 
```

> ## Motivation

Reviewer: Probably this chapter should be reformulated a bit

For the **Lido protocol**, TW support means a substantial reduction in trust assumptions toward Node Operators and Oracles. It unlocks the following capabilities:

> - **Permissionless Staking Modules**, such as CSM, where ETH cannot be "held hostage" even if the operator acts maliciously;

Reviewer: ```suggestion
- **Permissionless Staking Modules**, such as CSM, where ETH cannot be "held hostage" even if the operator misbehaves or significantly underperforms;
```

> - **Dual Governance**, allowing stETH holders to safely and autonomously withdraw their funds before proposed protocol changes are finalized;

Reviewer: here said 'TW unlocks DG';

not really, to my understanding DG complements TW making it safer against governance takeover type of attacks

- A mechanism for **emergency validator exits**, in case of key loss;
- **Direct DAO interaction**, enabling the Lido DAO to request validator exits independently of Oracles.

## Specification

> ## 1. Definitions and Terminology

Reviewer: ```suggestion
### 1. Gloassary
```

mind the headline level needed here (triple `###`)

### **Validators Exit Bus (VEB)**

> An on-chain contract that serves as the central infrastructure for managing validator exit requests. It stores report hashes, emits exit events, and maintains data and tools that enables anyone to prove a validator was requested to exit. Unlike VEBO, it supports exit reports from a wide range of entities.

Reviewer: what are the `report hashes`? (it's not clear for the reader)

### **Validators Exit Bus Oracle (VEBO)**

> A component of the oracle system responsible for delivering exit request reports to the VEB on behalf of oracles. Its primary role is to initiate the exit of enough validators to efficiently fulfill pending withdrawal requests. It is a part of the VEB contract.

Reviewer: would mention what oracles (the existing committee and reference doc link)

---
lip: 30
title: Triggerable Withdrawals Framework
status: WIP
author: Raman Siamionau, Evgeniy Pirogov
discussions-to: TBD
created: 2025-05-21
updated: 2025-06-28
---

# Simple Summary

This proposal outlines the implementation of the Triggerable Withdrawals framework within the Lido protocol. The goal is to enable permissionless, secure, and verifiable validator exits initiated from the Execution Layer. This enhancement aims to:
- Improve the protocol’s fault tolerance
- Reduce trust assumptions on node operators
- Strengthen the foundation for permissionless staking within Lido

Currently, Lido relies on the Validator Exit Bus Oracle (VEBO) and places trust in Node Operators to initiate validator exits on CL. If a node operator fails to comply with the protocol’s policies, [penalties are applied](https://docs.lido.fi/guides/oracle-spec/penalties).

> In the new version of the framework, this mechanism remains in place. However, it adds a new capability: the ability to exit validators who have requested to exit without requiring Node Operator action. This reduces reliance on node operators and aligns with Lido’s move toward a more decentralized and trust-minimized protocol design.

Reviewer: ```suggestion
In the new version of the framework based on the recently adopted EIP-7002, this mechanism remains in place. However, it adds a new capability: withdrawing validators who have requested to exit without requiring Node Operator action. This reduces reliance on node operators and aligns with a more decentralized and trust-minimized Lido protocol design vision.
```

# Abstract

> The **Triggerable Withdrawals (TW)** mechanism is a critical extension of Lido’s architecture, enabling the initiation of validator exits (and partial withdrawals in future) from the EL side without relying on the involvement of a Node Operator from the CL side. TW is based on [EIP-7002](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7002.md), which addresses one of the key long-standing issues in delegated staking on Ethereum. Previously, stakers had to rely on the goodwill of Node Operators - either to pre-sign an exit message or to agree to process it in the future. This limitation is now removed: any party with access to a validator’s withdrawal credentials (non-`0x00` types) can initiate its exit directly via the Execution Layer. 

Reviewer: ```suggestion
The **Triggerable Withdrawals (TW)** mechanism is a critical extension of the Lido protocol’s architecture, enabling the initiation of validator exits (and partial withdrawals in future) from the EL side without relying on the involvement of a Node Operator from the CL side. TW is based on [EIP-7002](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7002.md), which addresses one of the long-standing issues in delegated staking on Ethereum. Previously, stakers had to rely on the goodwill of Node Operators—either to pre-sign an exit message or to agree to process it in the future. This limitation is now lifted: any party with access to a validator’s withdrawal credentials (non-`0x00` types) can initiate its exit directly via the Execution Layer. 
```

# Motivation
For the **Lido protocol**, TW support means a substantial reduction in trust assumptions toward Node Operators and Oracles. It unlocks the following capabilities:

- **Permissionless Staking Modules**, such as CSM, where ETH cannot be "held hostage" even if the operator misbehaves or significantly underperforms;

> - A mechanism for **emergency validator exits**, in case of key loss;

Reviewer: ```suggestion
- A mechanism for **emergency validator exits**, in case of key loss or a potential compromise event;
```

- **Direct DAO interaction**, enabling the Lido DAO to request validator exits independently of Oracles.

> - Enables permissionless exits for validators that have been requested to exit, preventing NOs from delaying withdrawal queue fulfillment.

Reviewer: ```suggestion
- Enables permissionless exits for validators that have been requested to exit, preventing NOs from delaying withdrawal requests fulfillment.
```

- [Withdrawal Vault](https://github.com/lidofinance/core/blob/feat/triggerable-exits/contracts/0.8.9/WithdrawalVault.sol)
    - [WithdrawalVaultEIP7002](https://github.com/lidofinance/core/blob/feat/triggerable-exits/contracts/0.8.9/WithdrawalVaultEIP7002.sol)
- [Staking Router](https://github.com/lidofinance/core/blob/feat/triggerable-exits/contracts/0.8.9/StakingRouter.sol)
- [Node Operator Registry](https://github.com/lidofinance/core/blob/feat/triggerable-exits/contracts/0.4.24/nos/NodeOperatorsRegistry.sol)
- [New] [Validator Exit Delay Verifier](https://github.com/lidofinance/core/blob/feat/triggerable-exits/contracts/0.8.25/ValidatorExitDelayVerifier.sol)
- [New] [Triggerable Withdrawals Gateway](https://github.com/lidofinance/core/blob/feat/triggerable-exits/contracts/0.8.9/TriggerableWithdrawalsGateway.sol)

Off-chain Components:

- [Lido Oracle](https://github.com/lidofinance/lido-oracle/pull/685)
- [Validator Ejector](https://github.com/lidofinance/validator-ejector)
- [New] Trigger Exits Bot
- [New] Validator Late Prover Bot

## 3. Architecture and Scenarios

> At a high level, the architecture is centered around the **VEB**, which serves as the main coordination point for validator exit requests. Exit reports can be submitted to the VEB by trusted entities such as oracles or governance, which then emit validator exit events. Once exit request events are emitted, Node Operators are expected to voluntarily initiate the validator exit.

Reviewer: it's said here that Oracles already can work with VEB, but later on, VEBO suggested as a path forward :thinking:

- Validating the roles and permissions of the caller;
- Enforcing rate limits for safety and protocol stability;
- Coordinating with the `Withdrawal Vault` to trigger validator exits via the Execution Layer (EL);

Once an exit event has been emitted via the VEB, **any participant** can submit a TWR through a permissionless interface provided by the VEB contract. VEB ensures that the validator was indeed requested to exit before passing the request to the TWG for execution.

Additionally, the architecture includes a **Validator Exit Delay Verifier**, a contract that allows anyone to report late validators—those that failed to begin exiting within the expected time window after an exit event was emitted. These reports are submitted on-chain and forwarded to the appropriate staking modules, which may apply penalties based on module-specific logic.

### 3.1 Report Validators to Exit

> This flow describes the **two-phase delivery model** for reporting validators to exit. In the first phase, Oracles - via consensus on a smart contract, [Easy Tracks](https://docs.lido.fi/guides/easy-track-guide/) or governance submit a hash of the exit requests data to the **VEB**. The report data includes a list of validators to be exited.

Reviewer: ```suggestion
This flow describes the **two-phase delivery model** for reporting validators to exit. In the first phase, Oracles - via consensus on a smart contract or a governance (possibly represented via [Easy Track](https://docs.lido.fi/guides/easy-track-guide/ factoeries) can submit a hash of the exit requests data to the **VEB**. The report data includes a list of validators to be exited.
```

Reviewer: Easy Track **is** the governance, not an alternative mechanism!

The second phase involves revealing the actual report data and emitting exit events for the listed validators. When the data is revealed, the contract saves a timestamp linked to the data hash. This later allows any actor to prove that a specific validator was requested to exit, and when the corresponding event was emitted.

> Because report hashes can be submitted by more than just oracles, this system increases protocol resilience and moves Lido toward a more decentralized architecture.

Reviewer: I don't get why we make this conclusion

Author: Oracles is a form of centralization - it can block the protocol’s heartbeating by stoping reports. Even though the DAO can change oracle participants, the risk remains that the new participants might also stop reporting.

However, with the current design, the DAO can submit a report to the VEB directly. This process is independent of any external actor and relies solely on the DAO’s will.

The second part with decentralization is indeed not very obvious. Will remove it.

- The provided exit data hash exists and matches the hash stored in the VEB.
- The corresponding data has already been revealed.

Once validated, the request is forwarded to the **TWG** — the central contract handling all triggerable exit requests. The **TWG**:

- Enforces frame-level rate limits.
- Forwards the request to the **Withdrawal Vault**, which interacts with the **EIP-7002 precompile** to trigger the actual exit.
- Notifies the associated staking module.
- Refunds any extra value provided.

---

#### Submitting TWR through CSM

This flow mirrors the previous one, but instead of using validators that were requested to exit, **trusted entities** can directly initiate an exit **without any prior exit request event in the VEB**.

This direct exit flow is available to specific entities such as **CSM**, but due to its powerful nature, modules must adhere to strict internal checks before using it. These include:

- Verifying that the validator key belongs to the module.

> - Ensuring the key was deposited through a **DSM**.

Reviewer: ```suggestion
- Ensuring the key was deposited via Deposit Security Module (**DSM**).
```

    The maximum capacity of the quota. The available quota cannot exceed this value. With time, the quota is gradually replenished — up to this maximum.
2. `exitsPerFrame`
    Defines how much new quota becomes available with each new frame. This enables gradual and predictable replenishment of the quota over time.
3. `frameDuration` 
       The duration of each frame, in seconds, after which `exitsPerFrame` exits can be restored.

**Track Remaining Quota**

The system keeps `prevExitRequestsLimit` — the remaining quota after the last usage.
When a new request arrives, the system checks how much time has passed since `prevTimestamp`. It calculates how many full `frameDuration` intervals (frames) have elapsed.
For each full frame passed, `exitsPerFrame` units are restored. The updated quota is calculated as:`restoredQuota = prevExitRequestsLimit + framesPassed * exitsPerFrame`, but it is capped at `maxExitRequestsLimit`.
The request amount is subtracted from the restored quota. This new value becomes the `prevExitRequestsLimit` for future calculations.
The `prevTimestamp` is advanced by `framesPassed * frameDuration`, anchoring the system for the next round of quota restoration.

> ### Appendix B - Easy Tracks for VEB

Reviewer: ```suggestion
### Appendix B - Easy Track factories for VEB
```

> To simplify the exit request process for Node Operators (NOs) from the **Curated** and **sDVT Staking Modules**, **Easy Tracks** will be set up to facilitate exit requests.

Reviewer: ```suggestion
To simplify the exit request process for Node Operators (NOs) from the **Curated** and **sDVT Staking Modules**, **Easy Track** factories will be set up to facilitate exit requests.
```
